File: 0852-peak-index-in-a-mountain-array/0852-peak-index-in-a-mountain-array.cpp
class Solution {
public:
    int isHill(vector<int> &arr, int pos) {
        if (pos == 0) return -1;
        if (pos == arr.size() - 1 ) return 1;
                                           
        if (arr[pos-1] < arr[pos] && arr[pos] > arr[pos+1]) return 0;
        if (arr[pos-1] < arr[pos] && arr[pos] < arr[pos+1]) return -1;
        if (arr[pos-1] > arr[pos] && arr[pos] > arr[pos+1]) return 1;
        return 10;  // non-reachable
    }
    
    int peakIndexInMountainArray(vector<int>& arr) {
        int front = 0;
        int end = arr.size() - 1;
        
        while(front <= end) {
            int mid = (front + end) / 2;
            if (isHill(arr, mid) == 0) return mid;
            else if (isHill(arr, mid) == -1) front = mid + 1;
            else if (isHill(arr, mid) == 1) end = mid - 1;
        }
        
        return -1;  // non-reachable
    }
};

File: 0852-peak-index-in-a-mountain-array/0852-peak-index-in-a-mountain-array.kt
class Solution {
    fun peakIndexInMountainArray(arr: IntArray): Int {
        var left: Int = 0
        var right: Int = arr.size - 1
        var mid: Int = 0

        while (left <= right) {
            mid = left + (right - left) / 2
            val position = isPeek(arr, mid)
            when (position) {
                0 -> return mid
                1 -> right = mid - 1
                -1 -> left = mid + 1
            }
        }

        return mid
    }

    fun isPeek(arr: IntArray, index: Int): Int {
        if (index == 0) return -1
        if (index == arr.size - 1) return 1

        if (arr[index] > arr[index - 1] && arr[index] > arr[index + 1]) return 0
        if (arr[index] < arr[index - 1]) return 1
        return -1
    }
}

